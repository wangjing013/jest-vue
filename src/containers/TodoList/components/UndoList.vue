<template>
  <div class="undo-list-container">
    <div>
      代办列表: <span data-test="count">{{ count }}</span>
    </div>
    <ul class="undo-list">
      <li data-test="item" v-for="(item, index) of list" :key="index">
        {{ item }}
        <span
          data-test="delete-button"
          @click="
            () => {
              handerDelete(index)
            }
          "
          >-</span
        >
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  props: {
    list: {
      type: Array,
      default: () => {
        return []
      }
    }
  },
  computed: {
    count() {
      return this.list.length
    }
  },
  methods: {
    handerDelete(index) {
      this.$emit('delete', index)
    }
  }
}
</script>

<style></style>
