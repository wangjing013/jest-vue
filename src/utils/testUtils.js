export const findTestWarpper = (warpper, selector) => {
  return warpper.findAll(`[data-test="${selector}"]`)
}
