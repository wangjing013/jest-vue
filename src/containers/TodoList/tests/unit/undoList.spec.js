import { shallowMount } from '@vue/test-utils'
import UndoList from '../../components/UndoList.vue'
import { findTestWarpper } from '@/utils/testUtils'

it('UndoList 参数为 [] 时, 对应 count 应该为 0, 列表无内容', () => {
  const warpper = shallowMount(UndoList, {
    propsData: {
      list: []
    }
  })
  expect(warpper.vm.list).toEqual([])
  // expect(warpper.vm.count).toBe(0) // 写法一
  const count = findTestWarpper(warpper, 'count')
  const items = findTestWarpper(warpper, 'item')
  expect(count.at(0).text()).toBe('0') // 写法二
  expect(items.length).toBe(0)
})

it('UndoList 参数为 [1,2,3] 时, 对应 count 应该为 3, 列表有内容,且存在删除按钮', () => {
  const warpper = shallowMount(UndoList, {
    propsData: {
      list: [1, 2, 3]
    }
  })
  // expect(warpper.vm.count).toBe(0) // 写法一
  const count = findTestWarpper(warpper, 'count')
  const items = findTestWarpper(warpper, 'item')
  const deleteButtons = findTestWarpper(warpper, 'delete-button')
  expect(count.at(0).text()).toBe('3') // 写法二
  expect(items.length).toBe(3)
  expect(deleteButtons.length).toBe(3)
})

it('UndoList 当点击删除按钮时对外触发删除事件', () => {
  const warpper = shallowMount(UndoList, {
    propsData: {
      list: [1, 2, 3]
    }
  })
  const deleteButton = findTestWarpper(warpper, 'delete-button').at(1)
  deleteButton.trigger('click')
  // 对外触发事件
  expect(warpper.emitted().delete).toBeTruthy()
  // 事件参数
  expect(warpper.emitted().delete[0][0]).toBe(1)
})
