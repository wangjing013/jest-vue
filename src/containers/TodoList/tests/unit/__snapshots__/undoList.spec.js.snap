// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList 参数为 [] 时, 对应 count 应该为 0, 列表无内容 1`] = `
<div class="undo-list-container">
  <div>
    代办列表: <span></span></div>
  <ul data-test="undo-list" class="undo-list"></ul>
</div>
`;

exports[`header 样式添加快照 1`] = `
<div class="header">
  <div class="content">
    TodoList
    <input data-test="input" placeholder="请输入代办事项" class="input">
  </div>
</div>
`;
